elastic/kibana
# [Stack management apps] Deprecate "enabled" Kibana setting

In `8.0` we deprecate the `enabled` Kibana settings for the apps. As we want to offer an alternative way for our users to disable the UI we had 2 options: add the stack management app to Features control or use another Kibana setting (`ui.enabled`) for that purpose. The former would be the ideal way as it is the easiest and can be done through our UI. Unfortunately the Features control does not support yet disabling apps that depend on ES privileges so the security team has [opened an issue to solve that](https://github.com/elastic/kibana/issues/114699).

In the meantime we are going to rely on the `ui.enabled` Kibana setting to let our users disable the UI of our apps.
Some of our apps already use this setting (CCR, ILM, license management, remote clusters) so for those ones I've only added the deprecation notice for UA. For the rest of our apps I have added the `ui.enabled` setting to the plugin config. 

## Notes on implementation

As I mentioned several times I like the idea of having the same code base in `master` and `7.x`. The reasons are:

* It simplifies the DX. I _always_ work on `master` and can test different behaviours on `7.x` by changing 1 constant
* It avoids regression or forgetting bringing back code from `7.x` to `master` (when working against `7.x`), like we had to do in https://github.com/elastic/kibana/pull/113672 and https://github.com/elastic/kibana/pull/114966
* It is more efficient. There is no need to create multiple PRs, working on different master branch and bootstrap Kibana (which we all know can take above 20 min between `yarn kbn bootstrap` and `yarn start`)
* It avoids merge conflict and (again) regressions. If we work on `7.x` and modify a component we won't ever be able to automatically backport that component from `master` after that. We'll need to manually backport the change **and** remember why `7.x` is different and be sure to keep the change in place. Sure tests should have our back but it is not a fun process. Specially if the component is big and is not in our domain expertise.
* It is explicit. The code is the documentation. No need to add bunch of comments here and there. I read the code, I know what behaviour is in `7.x` and what is in other majors.
* It seems that an approach like this one is a step in the direction of "make it minor" where we might have different "breaking" behaviour between 2 minors that our apps need to handle gracefully for our users. We want to be able to do that without juggling between branches (and again avoid regression in the process).

## How to test

**Note** I have aligned all apps to the same naming convention (using the `MAJOR_VERSION` const) and file structure (expose that constant from the `common` folder, handle all the plugin config from inside the `./server/config.ts` file).

* Add the `enabled` (e.g. `xpack.index_management.enabled: false`) setting in `kibana.yml` and check that the server throws an error
* Change that setting to `ui.enabled`. The server restarts and the UI is not there
* Change the `MAJOR_VERSION` const (in the `common` folder) to `7.16.0` (the minor does not matter).
* The app UI should still not show
* Change back the `ui.enabled` setting to `enabled`. This time the server should not throw (we are simulating being on 7.16.0) and the UI should be disabled
* Open `x-pack/plugins/upgrade_assistant/public/plugin.ts` and change L64 (replace `readonly` with `false`)
* Open Upgrade assistant UI. You should see the deprecation for the `enabled` setting.
* Repeat for all the apps.

### Test Console

Console expose a URL locator to other apps to link to it. If console UI is disabled the apps linking to it should not display the console button.

* Set `console.ui.enabled: false` in the `kibana.yml`
* Navigate to the home page. You should not see "Dev tools" in the right side menu under "Management"

<img width="1237" alt="Screenshot 2021-10-14 at 08 00 34" src="https://user-images.githubusercontent.com/2854616/137267575-9b1fbfaf-60f3-4e8d-bbd6-8b7ee1e41463.png">

* From home navigate to "Analytics". Again here you should not see the "Dev tools" button

<img width="1263" alt="Screenshot 2021-10-14 at 08 02 15" src="https://user-images.githubusercontent.com/2854616/137267762-9069eeb7-c8e9-47ea-b860-5c506c2176ec.png">

* Open `src/plugins/kibana_overview/public/components/overview/overview.tsx` and change L212 (`solutions.length ?` to `false`).
* You should now see in the home page the "Manage your data" section. Make sure that "Interact with the Elasticsearch API" is not there.

<img width="1123" alt="Screenshot 2021-10-14 at 08 05 57" src="https://user-images.githubusercontent.com/2854616/137268281-1b28686f-d077-4bc9-b9a5-dcfe26d36885.png">

* Navigate to a dashboard and inspect one of the visualization

<img width="479" alt="Screenshot 2021-10-14 at 08 10 37" src="https://user-images.githubusercontent.com/2854616/137269349-f9b57b54-1b7a-4118-85ad-7538dce76cdc.png">

* Select "View: requests" the dropdown > Under the the "Request" tab you should not see the "Open in console" button

<img width="851" alt="Screenshot 2021-10-14 at 08 12 10" src="https://user-images.githubusercontent.com/2854616/137269523-24413d8e-da96-43cc-a15b-150e542e64a2.png">

## Improve console.proxyConfig and `console.proxyFilter` warning

As we are very close to FF I've added https://github.com/elastic/kibana/pull/114768/commits/6813c6aff03e8ac7c5f9e440f10d416df7816a07 to improve the deprecation message for the `console.proxyConfig` and `console.proxyFilter` settings.


<img width="1023" alt="Screenshot 2021-10-18 at 14 05 34" src="https://user-images.githubusercontent.com/2854616/137741588-e99e8d5d-500b-4c2f-9177-86d4e682991a.png">

<img width="809" alt="Screenshot 2021-10-18 at 14 05 19" src="https://user-images.githubusercontent.com/2854616/137741618-fccd034f-aaea-4176-b267-ec23e3e06d13.png">

Fixes https://github.com/elastic/kibana/issues/112605
Fixes https://github.com/elastic/kibana/issues/113810
